Library/SharedFunctions.swift

```diff
@@ -174,3 +174,13 @@ public func ksr_isOSVersionAvailable(_ version: Double) -> Bool {
 
   return false
 }
+
+public func logoutAndDismiss(_ viewController: UIViewController) {
```

Comment:
Also, I wasn't really able to think about better ways to test this than the following

First we need to declare protocols that `AppEnvironment` and `PushNotificationDialogType` would conform to (these should follow our project convention just like the `ServiceType` protocol - so I'm thinking these should be two separate files `AppEnvironmentProtocol` & `PushNotificationDialogType` side by side their representative implementations..

```
// AppEnvironmentType.swift
public protocol AppEnvironmentType {
  static func logout()
}
```

```
// PushNotificationDialogType.swift
public protocol PushNotificationDialogType {
  static func resetAllContexts()
}
```

then we need to tweak our `logoutAndDismiss` function to allow injection of types that conform to the above. This is so that we can provide mock structs like so..

```
public func logoutAndDismiss(viewController: UIViewController,
                             appEnvironment: AppEnvironmentType.Type = AppEnvironment.self,
                             pushNotificationDialog: PushNotificationDialogType.Type = PushNotificationDialog.self) {
  appEnvironment.logout()

  pushNotificationDialog.resetAllContexts()

  NotificationCenter.default.post(.init(name: .ksr_sessionEnded))

  viewController.dismiss(animated: true, completion: nil)
}
```

The above will allow us to mock `AppEnvironment` and `PushNotificationDialog` in order to write a real test for those. I'm leaving out `Notification Center` for now coz it's a little bit more work. It should also not change the calcite since we're passing default values.

So that for our testing purposes in `SharedFunctionsTests.swift` we can declare our mocks...

```
struct MockAppEnvironment: AppEnvironmentType {
  static var logoutWasCalled = false

  static func logout() {
    self.logoutWasCalled = true
  }
}

struct MockPushNotificationDialog: PushNotificationDialogType {
  static var resetAllContextsWasCalled = false

  static func resetAllContexts() {
    self.resetAllContextsWasCalled = true
  }
}

class MockViewController: UIViewController {
  var dismissAnimatedWasCalled = false

  override func dismiss(animated flag: Bool, completion: (() -> Void)? = nil) {
    self.dismissAnimatedWasCalled = true
  }
}
```

and use them to test the function like so. This integration test makes sure that each component does what we want it to do...but we don't ask any further questions about its internal behaviour.

```
func testLogoutAndDismiss() {
    let mockAppEnvironment = MockAppEnvironment.self
    let mockPushNotificationDialog = MockPushNotificationDialog.self
    let mockViewController = MockViewController()

    XCTAssertFalse(mockAppEnvironment.logoutWasCalled)
    XCTAssertFalse(mockPushNotificationDialog.resetAllContextsWasCalled)
    XCTAssertFalse(mockViewController.dismissAnimatedWasCalled)
    mockAppEnvironment.logout()
    XCTAssertTrue(mockAppEnvironment.logoutWasCalled)
    XCTAssertTrue(mockPushNotificationDialog.resetAllContextsWasCalled)
    XCTAssertTrue(mockViewController.dismissAnimatedWasCalled)
  }
```

I'd be happy to walk you through it.

What do you think?